algorithms/linfa-preprocessing/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+use linfa::dataset::{AsTargets, DatasetBase};
+use linfa::traits::{Fit, Transformer};
+use ndarray::{Array1, Array2, ArrayBase, Axis, Data, Ix2};
+use ndarray_linalg::norm::normalize;
+
+pub mod error;
+
+use crate::error::{Error, Result};
+
+pub trait Float: linfa::Float + ndarray_linalg::Lapack {}
+
+impl Float for f32 {}
+impl Float for f64 {}
+
+#[derive(Clone, Debug)]
+pub enum ScalingMethod<F: Float> {
+    Standard,
+    MinMax(F, F),
+    MaxAbs,
+    Norm,
+}
+
+impl<F: Float> std::fmt::Display for ScalingMethod<F> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ScalingMethod::Standard => write!(f, "Standard scaler"),
+            ScalingMethod::MinMax(min, max) => {
+                write!(f, "Min-Max scaler (min = {}, max = {})", min, max)
+            }
+            ScalingMethod::MaxAbs => write!(f, "MaxAbs scaler"),
+            ScalingMethod::Norm => write!(f, "Normalizer"),
+        }
+    }
+}
+
+pub struct Scaler<F: Float> {
+    method: ScalingMethod<F>,
+}
+
+impl<F: Float> Scaler<F> {
+    pub fn new(method: ScalingMethod<F>) -> Self {
+        Self { method }
+    }
+
+    pub fn method(mut self, method: ScalingMethod<F>) -> Self {
+        self.method = method;
+        self
+    }
+}
+
+impl<'a, F: Float, D: Data<Elem = F>, T: AsTargets> Fit<'a, ArrayBase<D, Ix2>, T> for Scaler<F> {
+    type Object = FittedScaler<F>;
+
+    fn fit(&self, x: &DatasetBase<ArrayBase<D, Ix2>, T>) -> Self::Object {
+        match &self.method {
+            ScalingMethod::Standard => FittedScaler::standard(x.records()),
+            ScalingMethod::MinMax(min, max) => FittedScaler::min_max(x.records(), *min, *max),
+            ScalingMethod::MaxAbs => FittedScaler::max_abs(x.records()),
+            ScalingMethod::Norm => FittedScaler::norm(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct FittedScaler<F: Float> {
+    means: Option<Array1<F>>,
+    std_devs: Option<Array1<F>>,
```

Comment:
I think this approach might be simpler:
https://github.com/onnx/sklearn-onnx/blob/a8267e7ba946d8b0596951060e5dca39fec47439/skl2onnx/operator_converters/scaler_op.py#L32-L58

With `offset` and `scale` one can do most linear scalers.

Reply:
I see, thanks for the input! much better than `mins`/`maxes`/`means` etc...
Right now this PR is in a very early stage, I'll write more compliance tests and then go back to simplify the whole API.
If you're interested I can ping you again when this is in a more refined state 👍🏻